src/serde/deserialize_utils.rs

```diff
@@ -0,0 +1,359 @@
+        return hex;
+    }
+    hex
+}
+
+pub fn parse_dereference(value: &str) -> Result<ValueAddress, ReferenceParseError> {
+    let splitted: Vec<&str> = value.split(" + ").collect();
+
+    match splitted.len() {
+        1 => parse_dereference_no_offsets(splitted),
+        2 => parse_dereference_with_one_offset(splitted),
+        3 => parse_dereference_with_two_offsets(splitted),
+        _ => Err(ReferenceParseError::InvalidStringError(String::from(value))),
+    }
+}
+// parse string values of format `[cast(reg, *felt)]`
+fn parse_dereference_no_offsets(
+    splitted_value_str: Vec<&str>,
+) -> Result<ValueAddress, ReferenceParseError> {
+    let str_tmp: Vec<&str> = splitted_value_str[0].split(',').collect();
+
+    let register = match str_tmp[0].split('(').collect::<Vec<_>>()[1] {
+        "ap" => Some(Register::AP),
+        "fp" => Some(Register::FP),
+        _ => None,
+    };
+
+    Ok(ValueAddress {
+        register,
+        offset1: 0,
+        offset2: 0,
+        immediate: None,
+        dereference: true,
+    })
+}
+
+// parse string values of format `[cast(reg + offset1, *felt)]`
+fn parse_dereference_with_one_offset(
+    splitted_value_str: Vec<&str>,
+) -> Result<ValueAddress, ReferenceParseError> {
+    let register = match splitted_value_str[0].split('(').collect::<Vec<&str>>()[1] {
+        "ap" => Some(Register::AP),
+        "fp" => Some(Register::FP),
+        _ => None,
+    };
+
+    let mut offset1_str = splitted_value_str[1].split(',').collect::<Vec<_>>()[0].to_string();
+    offset1_str.retain(|c| !r#"()]"#.contains(c));
+
+    let offset1: i32 = match offset1_str.parse() {
+        Ok(offset1) => offset1,
+        // for the moment, references with values that overflow i32 are not important, they are just dummy references.
+        Err(e) => match e.kind() {
+            IntErrorKind::PosOverflow => 0,
+            _ => return Err(ReferenceParseError::IntError(e)),
+        },
+    };
+
+    Ok(ValueAddress {
+        register,
+        offset1,
+        offset2: 0,
+        immediate: None,
+        dereference: true,
+    })
+}
+
+// parse string values of format `[cast([reg + offset1] + offset2, *felt)]`
+fn parse_dereference_with_two_offsets(
+    splitted_value_str: Vec<&str>,
+) -> Result<ValueAddress, ReferenceParseError> {
+    let register = match splitted_value_str[0].split('[').collect::<Vec<&str>>()[2] {
+        "ap" => Some(Register::AP),
+        "fp" => Some(Register::FP),
+        _ => None,
+    };
+
+    let mut offset1_str = splitted_value_str[1].to_string();
+    offset1_str.retain(|c| !r#"()]"#.contains(c));
+
+    let offset1: i32 = match offset1_str.parse() {
+        Ok(offset1) => offset1,
+        // for the moment, references with values that overflow i32 are not important, they are just dummy references.
+        Err(e) => match e.kind() {
+            IntErrorKind::PosOverflow => 0,
+            _ => return Err(ReferenceParseError::IntError(e)),
+        },
+    };
```

Comment:
Similarly:
```suggestion
) -> Result<ValueAddress, ReferenceParseError> {
    let mut deref = parse_dereference_with_one_offset(splitted_value_str);
```
And modify `deref.offset2` at the end.